### PR TITLE
chore: Cosmos SDK 0.44.* series upgrade for Osmosis

### PR DESCRIPTION
**Pull request by faddat**
Closes: #419 

## Description

This is ready for review.  One thing that'll need review especially is the decimal math stuff in the faddat/basecoin cosmos sdk.  I am working to get that up to 0.44.2 and when that's done, the sdk stuff will need review also. 

Thanks!

Tracking checklist:

- [x] Update my existing cosmos-sdk code to v0.44.2
- [x] Merge the  v0.44.2 cosmos sdk into osmosis-sdk
- [x] Get a proper Cosmos-sdk [pushed to osmosis-labs/cosmos-sdk](https://github.com/osmosis-labs/cosmos-sdk/tree/faddat/0442-osmo) instead of faddat/basecoin
- [x] ~~Cherry pick~~ (merge?) individual commits on [Osmo-v2-upgrade](https://github.com/osmosis-labs/cosmos-sdk) into 0.44.2 branch <- maybe-complete. 
- [x] create a 045-osmo branch and bring it current with sdk mainline-- this branch will become mainline-osmo once 045 is cut 
- [x] Update to 0.44.3
- [x] Make tests work again
- [ ] Use this as a jumping off point for putting #467 into action
- [ ] IBC-go/v2 readiness

NB:  This has also added & enabled authz and feegrant.  It should also ease some of @sunnya97's work on ethereum address formats. 

### Big big Goal
Osmo is IBC yo.

oh, and I hate, hate hate Marshaler. 

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by sunnya97** [comment]
@faddat what kind of performance upgrades do we get from an export state upgrade?

**Comment by faddat** [comment]
@sunnya97 the performance improvements tend to be quite dramatic.  

That said, we surely cannot lose the existing IBC channels, so I will look into other ways to roll.  

I'm also going to look into weather it is realistic to include this in v5.x

**Comment by ValarDragon** [comment]
I think its realistic to include this in v5.x!

We made a branch for the changes by cherry-picking every commit manually instead of the rebase hell. (Spent awhile trying to figure out how to double check everything in a rebase and its super hard)

I think we can use `osmosis-labs/cosmos-sdk branch v0.44.3x-osmo-v5` in the go.mod.

If your fine with it, can I just push to this branch directly? I think it should take some minor changes before this is merge ready!

File: app/app.go [hunk 1]
```diff
@@ -304,32 +309,34 @@ func NewOsmosisApp(
...
+				app.LockupKeeper.ClearAllAccumulationStores(ctx)
 
-			// reset all lock and references
-			if err := app.LockupKeeper.ResetAllLocks(ctx, locks); err != nil {
-				panic(err)
-			}
+				// reset all lock and references
+				if err := app.LockupKeeper.ResetAllLocks(ctx, locks); err != nil {
+					panic(err)
+				}
```
Comment (ValarDragon) [inline app/app.go:330]: I think we should un-comment this part out. I would like all handlers to still be defined personally. (Though maybe we move it to a new file and comment out the actual code.)

With the handlers, we can do things like https://github.com/cosmos/cosmos-sdk/pull/10407

File: cmd/osmosisd/cmd/airdrop.go [hunk 2]
```diff
@@ -60,188 +49,188 @@ func setCosmosBech32Prefixes() {
...
+// 			}
+
+// 			totalOsmoBalance := sdk.NewInt(0)
+// 			onePointFive := sdk.MustNewDecFromStr("1.5")
+
+// 			for address, acc := range snapshotAccs {
+// 				allAtoms := acc.AtomBalance.ToDec()
+
+// 				acc.AtomOwnershipPercent = allAtoms.QuoInt(totalAtomBalance)
+
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.go:167]: If we have the airdrop.history file, can we delete this file entirely?

File: cmd/osmosisd/cmd/airdrop.history [hunk 3]
```diff
@@ -0,0 +1,253 @@
+/*
+
+Preserved for historical purposes
+package cmd
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/airdrop.history:4]: Can we rename this file `airdrop.go.history`

Comment (faddat) [reply cmd/osmosisd/cmd/airdrop.history:4]: awesome

File: x/pool-incentives/keeper/keeper.go [hunk 4]
```diff
@@ -140,12 +140,12 @@ func (k Keeper) GetLockableDurations(ctx sdk.Context) []time.Duration {
...
 
 	return info.LockableDurations
 }
 
-func (k Keeper) GetAllGauges(ctx sdk.Context) []incentivestypes.Gauge {
-	gauges := k.incentivesKeeper.GetGauges(ctx)
-	return gauges
+func (k Keeper) ExportGenesis(ctx sdk.Context) interface{} {
+	fmt.Println("you have hit a very silly placeholder, smartly suggested by the IDE")
+	return nil
```
Comment (ValarDragon) [inline x/pool-incentives/keeper/keeper.go:150]: Something odd happened here

Comment (faddat) [reply x/pool-incentives/keeper/keeper.go:155]: heh, let me have a look-see

File: app/app.go [hunk 5]
```diff
@@ -288,8 +288,13 @@ func NewOsmosisApp(
 		appCodec, keys[authtypes.StoreKey], app.GetSubspace(authtypes.ModuleName), authtypes.ProtoBaseAccount, maccPerms,
 	)
 	app.BankKeeper = bankkeeper.NewBaseKeeper(
-		appCodec, keys[banktypes.StoreKey], app.AccountKeeper, app.GetSubspace(banktypes.ModuleName), app.BlockedAddrs(),
+		appCodec,
+		keys[banktypes.StoreKey],
+		app.AccountKeeper,
+		app.GetSubspace(banktypes.ModuleName),
+		app.ModuleAccountAddrs(),
```
Comment (ValarDragon) [inline app/app.go:295]: This got changed from `app.BlockedAddrs()` to `app.ModuleAccountAddrs()`.

We need to undo this change

File: app/app.go [hunk 6]
```diff
@@ -395,10 +412,11 @@ func NewOsmosisApp(
 	// register the proposal types
 	govRouter := govtypes.NewRouter()
 	govRouter.AddRoute(govtypes.RouterKey, govtypes.ProposalHandler).
-		AddRoute(paramproposal.RouterKey, params.NewParamChangeProposalHandler(app.ParamsKeeper)).
+		AddRoute(paramstypes.QuerierRoute, params.NewParamChangeProposalHandler(app.ParamsKeeper)).
```
Comment (ValarDragon) [inline app/app.go:415]: TODO: Double check this

File: app/app.go [hunk 7]
```diff
@@ -591,8 +609,8 @@ func NewOsmosisApp(
 		// that in-memory capabilities get regenerated on app restart.
 		// Note that since this reads from the store, we can only perform it when
 		// `loadLatest` is set to true.
-		ctx := app.BaseApp.NewUncachedContext(true, tmproto.Header{})
-		app.CapabilityKeeper.InitializeAndSeal(ctx)
+		// ctx := app.BaseApp.NewUncachedContext(true, tmproto.Header{})
+		app.CapabilityKeeper.Seal()
```
Comment (ValarDragon) [inline app/app.go:613]: TODO: Double check

File: app/config.go [hunk 8]
```diff
@@ -28,26 +29,30 @@ func DefaultConfig() network.Config {
...
-		StakingTokens:   sdk.TokensFromConsensusPower(500),
-		BondedTokens:    sdk.TokensFromConsensusPower(100),
-		PruningStrategy: storetypes.PruningOptionNothing,
-		CleanupDir:      true,
-		SigningAlgo:     string(hd.Secp256k1Type),
-		KeyringOptions:  []keyring.Option{},
+		AppConstructor:    NewAppConstructor(encCfg),
+		GenesisState:      ModuleBasics.DefaultGenesis(encCfg.Marshaler),
+		TimeoutCommit:     2 * time.Second,
+		ChainID:           "osmosis-code-test",
```
Comment (ValarDragon) [inline app/config.go:35]: TODO: Double check this still gets set correctly in `client.toml` after `osmosisd init`

File: x/claim/keeper/hooks.go [hunk 9]
```diff
@@ -53,6 +53,14 @@ func (k Keeper) Hooks() Hooks {
 }
 
 // gamm hooks
```
Comment (ValarDragon) [inline x/claim/keeper/hooks.go:55]:
```suggestion
// gamm and governance hooks
```

File: x/claim/keeper/hooks.go [hunk 10]
```diff
@@ -53,6 +53,14 @@ func (k Keeper) Hooks() Hooks {
...
+func (h Hooks) AfterProposalFailedMinDeposit(ctx sdk.Context, proposalId uint64) {
+
+}
+
+func (h Hooks) AfterProposalVotingPeriodEnded(ctx sdk.Context, proposalId uint64) {
+
+}
+
```
Comment (ValarDragon) [inline x/claim/keeper/hooks.go:63]: Not super sure why these are added, should they be deleted?

**Comment by ValarDragon** [comment]
All tests now working. Now need to fix simulation

File: app/app.go [hunk 11]
```diff
@@ -304,43 +309,55 @@ func NewOsmosisApp(
...
+
+			// // reset all lock and references
+			// if err := app.LockupKeeper.ResetAllLocks(ctx, locks); err != nil {
+			// 	panic(err)
+			// }
+
+			// // configure upgrade for gamm module's pool creation fee param add
+			// app.GAMMKeeper.SetParams(ctx, gammtypes.NewParams(sdk.Coins{sdk.NewInt64Coin("uosmo", 1)})) // 1 uOSMO
 
 			prop12(ctx, app)
```
Comment (sunnya97) [inline app/app.go:341]: Why is this not commented out?

Comment (ValarDragon) [reply app/app.go]: Thanks. Done. The reason it wasn't initially was just due to having to comment out the related test as well. Thats now done!

File: app/app.go (hunk 6 above)
Comment (sunnya97) [reply app/app.go:414]: Don't see why this is swapped to QuerierRoute.  Should still just use RouterKey

See upstream:  https://github.com/cosmos/cosmos-sdk/blob/master/simapp/app.go#L293

File: app/app.go (hunk 7 above)
Comment (sunnya97) [reply app/app.go:614]: The can be moved out of the loadLatest condition.

File: app/app.go (hunk 6 above)
Comment (ValarDragon) [reply app/app.go]: Undid that, wasn't clear to me either on reviewing the changes

File: app/ante.go [hunk 12]
```diff
@@ -35,10 +35,9 @@ func NewAnteHandler(
 		ante.TxTimeoutHeightDecorator{},
 		ante.NewValidateMemoDecorator(ak),
 		ante.NewConsumeGasForTxSizeDecorator(ak),
-		ante.NewRejectFeeGranterDecorator(),
 		ante.NewSetPubKeyDecorator(ak), // SetPubKeyDecorator must be called before all signature verification decorators
 		ante.NewValidateSigCountDecorator(ak),
-		ante.NewDeductFeeDecorator(ak, bankKeeper),
+		ante.NewDeductFeeDecorator(ak, bankKeeper, nil),
```
Comment (sunnya97) [inline app/ante.go:40]: Upstream, DeductFee happens before SetPubKey.  We should probably copy for sanity sake

File: x/claim/keeper/hooks.go (hunk 10 above)
Comment (sunnya97) [reply x/claim/keeper/hooks.go:63]: Yeah, if they're not needed in the claims module, we should remove.

Comment (ValarDragon) [reply x/claim/keeper/hooks.go]: Ah its for interface matching Claims `Hooks` as GovHooks. Probably not needed, but we're already adding extra hooks for gamm as it is. This can be resolved later.

File: x/lockup/client/cli/cli_test.go [hunk 13]
```diff
@@ -221,10 +226,16 @@ func (s *IntegrationTestSuite) TestNewBeginUnlockPeriodLockCmd() {
...
+		sdk.NewCoins(sdk.NewCoin(s.cfg.BondDenom, sdk.NewInt(200))),
+		"1s")
 	s.Require().NoError(err)
-	s.Require().NoError(clientCtx.JSONMarshaler.UnmarshalJSON(out.Bytes(), &txResp), out.String())
-	lockID := txResp.Logs[0].Events[0].Attributes[0].Value
+	s.Require().NoError(clientCtx.JSONCodec.UnmarshalJSON(out.Bytes(), &txResp), out.String())
+	// This is a hardcoded path in the events to get the lockID
+	// this is incredibly brittle...
+	// fmt.Println(txResp.Logs[0])
+	lockID := txResp.Logs[0].Events[2].Attributes[0].Value
```
Comment (ValarDragon) [inline x/lockup/client/cli/cli_test.go:238]: Need to open an issue for this

File: x/lockup/keeper/gas_test.go [hunk 14]
```diff
@@ -70,6 +70,6 @@ func (suite *KeeperTestSuite) TestRepeatedLockTokensDistinctDurationGas() {
 
 	avgGas, maxGas := suite.measureAvgAndMaxLockGas(totalNumLocks, defaultAddr, coinsFn, durFn)
 	fmt.Printf("test deets: total locks created %d\n", totalNumLocks)
-	suite.Assert().EqualValues(110725, int(avgGas), "average gas / lock")
-	suite.Assert().EqualValues(231313, int(maxGas), "max gas / lock")
+	suite.Assert().EqualValues(122316, int(avgGas), "average gas / lock")
+	suite.Assert().EqualValues(242903, int(maxGas), "max gas / lock")
```
Comment (ValarDragon) [inline x/lockup/keeper/gas_test.go:74]: These constants are suspected to have gone up due to FundAccount instead of SetBalances.

**Review by ValarDragon — APPROVED** [review]
Everything looks good to me here, aside from my outstanding comments

File: app/app.go (hunk 7 above)
Comment (ValarDragon) [reply app/app.go]: Done, following gaia / regen's example

File: app/config.go (hunk 8 above)
Comment (ValarDragon) [reply app/config.go:35]: chain id is set under the logic:

"chainID, _ := cmd.Flags().GetString(flags.FlagChainID)
			if chainID == "" {
				chainID = fmt.Sprintf("test-chain-%v", tmrand.Str(6))
			}"
			
as far as I can tell.

Comment (ValarDragon) [reply app/config.go:35]: Ah no, client.toml happens entirely separately. I'll file an issue for updating that.

**Review by sunnya97 — APPROVED** [review]

**Comment by ValarDragon** [comment]
Done, opening follow-up issues now!
